task.c

```diff
@@ -0,0 +1,120 @@
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+
+#include "task.h"
+#include "common.h"
+#include "sockets.h"
+
+/* Tasks are stored in a consecutive chunk of memory, the first
+ * sizeof(task_spec) bytes are arranged acording to the struct
+ * task_spec. Then there is an array of task_args of length
+ * (num_args + num_returns), and then follows the data of
+ * pass-by-value arguments of size args_value_size. The offsets in the
+ * task_arg.val are with respect to the end of the struct, i.e.
+ * with respect to the address &task_spec.ids[0]. */
+
+typedef struct {
+  /* Either ARG_BY_REF or ARG_BY_VAL. */
+  int8_t type;
+  union {
+    unique_id arg_id;
+    struct {
+      /* Offset where the data associated to this arg is located relative
+       * to &task_spec.ids[0]. */
+      ptrdiff_t offset;
+      int64_t length;
+    } val;
+  };
+} task_arg;
+
+struct task_spec_impl {
+  function_id func_id;
```

Comment:
Thanks for comments. Much clear now.
Now that I understand better, 
1) your initial name arg_value_offset was better than arg_value_curosr; Similarly, arg_idx is better than arg_curosr;
2) I believe ids[0] stores both arguments as values and ids, right? If yes, I'd split it into: returns[0] to args[0].